### PR TITLE
ENH: Add fit_update() to LinearGaussianBayesianNetwork

### LINKED ISSUE #2946: [ENH] Add fit_update() for LinearGaussianBayesianNetwork
## Context

This is a follow-up to #2807, which is adding Bayesian estimation
to `LinearGaussianBayesianNetwork.fit()`. Once that is implemented,
a natural next step would be supporting incremental learning with
`fit_update()`.

`DiscreteBayesianNetwork` already provides `fit_update()`, but it
works using Dirichlet pseudo-counts internally, which is specific
to the discrete case. That approach does not directly apply to
continuous networks.

For `LinearGaussianBayesianNetwork`, the conjugate prior is
Normal-Inverse-Gamma. Because of this, the update logic would need
to be implemented separately.

## What fit_update() would do

After fitting the model using `estimator="bayes"`, the posterior
parameters (μₙ, Λₙ, αₙ, bₙ) from that fit could be reused as the
prior for the next batch of data. This would allow updating the
model incrementally without refitting from scratch.

Example:

model.fit(df_batch1, estimator="bayes")
model.fit_update(df_batch2)
model.fit_update(df_batch3)

Under the Normal-Inverse-Gamma conjugacy, the result after each
update should be mathematically identical to fitting the model on
all data together.

## Implementation idea

The sufficient statistics from the previous Bayesian fit could be
stored inside each `LinearGaussianCPD`. When `fit_update()` is
called, it would read those statistics, apply the NIG update
equations with the new batch of data, and update the stored
parameters.

If `fit_update()` is called before a Bayesian fit, it should raise
a clear error.

## Relation to existing work

- Depends on #2807 being merged first
- Similar in spirit to `fit_update()` in `DiscreteBayesianNetwork`,
  but requires a separate implementation because the prior
  structure is different

I'd like to work on this if it can be assigned.

cc @ankurankan

### PR DESCRIPTION
**Pull request by sohamjadhav95**
# The following checklist is mandatory.

Your PR will be closed if you remove the checklist or do not answer the questions to a satisfactory level. Use of LLMs is **strictly forbidden** for any part of this checklist (including for improving language), and will result in a **ban** if we find any use of LLMs.

### Your checklist for this pull request

- [x] Have you followed all the steps from our [Contributing Guide](https://github.com/pgmpy/pgmpy/blob/dev/Contributing.md)?
- [x] Does the PR fully address the linked issue and is within its defined scope? If you are still working on the PR, mark it as draft.
- [x] Are all the GitHub Actions checks passing? If not, mark your PR as draft while you fix it.

Please answer the following questions:

- Did you use an LLM for any assistance with this PR? Please describe in **detail** (around a paragraph) how and what you used it for?  
Yes, I used Claude in this PR. My starting point was the issue discussion with @ankurankan, where the approach is to updating the joint Gaussian using the pooled mean/covariance formula and then re-deriving CPDs was already decided. I understood the high-level idea from that, but needed help understanding it into a correct and working implementation. I used Claude to better understand the pooled covariance formula and how CPD parameters (regression coefficients and noise variance) can be derived from a joint Gaussian. Once I was clear on the math, I used it to draft an initial version of `fit_update()` based on that approach, along with the previous design that stored sufficient statistics in the model and CPDs. After the review, when @ankurankan suggested removing stored state, I reworked the implementation using `to_joint_gaussian()` to recover the joint distribution on the fly. I used Claude here to reason about the implications of this change. In particular, it pointed out the slight covariance inflation due to unbiased CPD std values, which I then verified myself numerically by comparing it with the actual sample covariance. I also used Claude to help restructure the tests (moving them into `TestLGBNMethods` and compressing them), and for debugging CI issues such as pre-commit formatting and dependency-related failures. Throughout the process, I carefully reviewed all generated code, ran tests locally, and verified results numerically. I made sure I understood each part of the implementation and can fully explain the logic including the joint Gaussian reconstruction, pooled update, and CPD parameter.

- Are you able to fully explain your changes? We expect you to fully understand the algorithm and take full responsibility for any changes in this PR.
Yes. The current implementation is stateless. Instead of storing sample statistics in the model or CPDs, `fit_update()` reconstructs the prior joint Gaussian using `to_joint_gaussian()`, and then combines it with the new batch using the pooled mean and covariance formula. From the updated joint distribution, CPD parameters are re-derived using conditional Gaussian relationships: regression coefficients from cross-covariance with parent variables, intercept adjusted using means, and noise variance from the conditional covariance (with adjustment to match pgmpy’s unbiased std estimator). I fully understand this pipeline and can explain each step in detail.

- What steps have you taken to verify that the changes correctly address the issue? And what edge cases have you considered? Other than running tests, what else have you verified?  
I verified numerically that `fit(df1)` followed by `fit_update(df2)` produces results very close to `fit(df_all)`. Due to reconstruction from CPD std values, there is a small numerical difference, which I analyzed and confirmed is expected. I also tested multiple sequential updates to ensure consistency across batches. 

- Has the LLM added try-except blocks? They will need to be removed; any error handling must be explicit.
No. All error handling is done explicitly using Value Error.

- Have you used LLM for generating tests? They need to be compressed into a smaller number of tests without reducing coverage.  
Yes. I used an LLM to help restructure and compress the tests, and then manually verified that they cover the same cases as before.

### Issue number(s) that this pull request fixes
- Fixes #2946 

### List of changes to the codebase in this pull request
- `LinearGaussianCPD`: removed previously added stored statistics to keep CPDs stateless
- `LinearGaussianBayesianNetwork`: implemented `fit_update()` using joint Gaussian reconstruction instead of stored state
- Tests: reorganized into `TestLGBNMethods` and reduced to focused cases covering correctness, sequential updates, and error handling

### CODE REVIEW

**Comment by sohamjadhav95** [comment]
Hey @ankurankan, all 23 CI checks are passing including doctests on Python 3.10 and 3.14. The implementation follows the approach we discussed, updating the joint Gaussian directly and re-extracting CPD parameters from the updated joint.

Let me know if anything should be adjusted.

**Comment by ankurankan** [comment]
@sohamjadhav95 Why was the checklist removed?

**Comment by sohamjadhav95** [comment]
@ankurankan I admire my previous mistake, I'd removed the checklist while editing the PR description. I've restored it now and filled all the required sections.

File: pgmpy/factors/continuous/LinearGaussianCPD.py [hunk 1]
```diff
@@ -104,6 +109,11 @@ def copy(self):
             evidence=list(self.evidence),
         )
 
+        if self._n_samples is not None:
+            copy_cpd._n_samples = self._n_samples
+            copy_cpd._sample_mean = self._sample_mean.copy()
+            copy_cpd._sample_cov = self._sample_cov.copy()
```
Comment (ankurankan) [inline pgmpy/factors/continuous/LinearGaussianCPD.py]: I would suggest not storing these. Some of these features will come in the next refactor of the parameterization in the package. But it would be better to do that once we have carefully thought about what and how we want to store it. Also, these are not properties of the `LinearGassianCPD`, so should definitely not be here.

File: pgmpy/models/LinearGaussianBayesianNetwork.py [hunk 2]
```diff
@@ -122,6 +122,10 @@ def __init__(
             roles=roles,
         )
         self.cpds = []
+        # Sufficient statistics stored by fit() and used by fit_update()
+        self._n_samples = None
+        self._sample_mean = None
+        self._sample_cov = None
```
Comment (ankurankan) [inline pgmpy/models/LinearGaussianBayesianNetwork.py]: Remove these as well. You can compute these on the fly in the `fit_update` method. For `n_samples`, you can ask to user to provide that. Have a look at `DiscreteBN.fit_predict`. We should try to match the method signature here as well.

File: pgmpy/models/LinearGaussianBayesianNetwork.py [hunk 3]
```diff
@@ -857,6 +861,126 @@ def fit(
 
         # Step 3: Add the estimated CPDs to the model
         self.add_cpds(*cpds)
+
+        # Step 4: Store sufficient statistics for use in fit_update()
+        variables = list(nx.topological_sort(self))
+        self._n_samples = len(data)
+        self._sample_mean = data[variables].mean().values
+        self._sample_cov = data[variables].cov(ddof=0).values
```
Comment (ankurankan) [inline pgmpy/models/LinearGaussianBayesianNetwork.py]: Need to remove this too.

File: pgmpy/models/LinearGaussianBayesianNetwork.py [hunk 4]
```diff
@@ -857,6 +861,126 @@ def fit(
...
+                beta_coeffs = np.linalg.solve(cov_pp, cov_ip)
+                beta_intercept = mu_updated[i] - beta_coeffs @ mu_updated[p_idx]
+                sigma2_mle = cov_updated[i, i] - cov_ip @ beta_coeffs
+
+                beta = np.append([beta_intercept], beta_coeffs)
+                std = float(np.sqrt(max(sigma2_mle, 0) * n_total / (n_total - k)))
+
+            # Update the CPD in place
+            cpd.beta = beta
+            cpd.std = std
```
Comment (ankurankan) [inline pgmpy/models/LinearGaussianBayesianNetwork.py]: Return new objects; don't modify in place. Again, refer the `DiscreteBN.fit_update` method. This method should also behave similarly to keep it intuitive.

File: pgmpy/tests/test_models/test_LinearGaussianBayesianNetwork.py [hunk 5]
```diff
@@ -619,3 +619,75 @@ def test_simulate_missing_prob_invalid_probability(self):
 
         with self.assertRaises(ValueError):
             model.simulate(n_samples=10, missing_prob={"X1": 1.5})
+
+
+class TestLGBNFitUpdate(unittest.TestCase):
```
Comment (ankurankan) [inline pgmpy/tests/test_models/test_LinearGaussianBayesianNetwork.py]: Don't create a new class for this. Should go in the `TestLGBNMethods` class. Compress these into max 2-3 test cases.

**Review by ankurankan — CHANGES_REQUESTED** [review]

**Comment by sohamjadhav95** [comment]
Hey @ankurankan, I've made the requested changes:

- Removed `_n_samples`, `_sample_mean`, `_sample_cov` from both `LinearGaussianCPD` and `LinearGaussianBayesianNetwork`. The implementation is now fully stateless.
- Updated `fit_update()` to recover the joint distribution using `to_joint_gaussian()` and apply the pooled update. `n_prev_samples` is now an explicit parameter.
- CPDs are now created as new objects and added via `add_cpds()`, so no in-place modification of CPD attributes.
- Moved the tests into `TestLGBNMethods` and reduced them to a couple of focused cases.

One small note on the test precision: since `to_joint_gaussian()` reconstructs covariance from the (unbiased) CPD std values, the recovered covariance is slightly scaled compared to the original sample covariance. Because of this, `fit(df1) + fit_update(df2)` isn't exactly identical to `fit(df_all)`, but the results are very close. I've adjusted the tolerance and added a short comment in the test to explain this.

Let me know if you'd like to make any adjustments further.

**Comment by ankurankan** [comment]
@sohamjadhav95, I need you to update your PR checklist to describe how exactly you have used LLMs. For example, refer https://github.com/pgmpy/pgmpy/pull/2622. I am confident from both our conversation and this PR that this is LLM agent generated. If that is the case, it need to accurately described in the checklist. As stated in the checklist, any inconsistency will lead to a ban from the repo.

**Comment by sohamjadhav95** [comment]
I updated the checklist to more detailed way how I used LLMs.

**Comment by ankurankan** [comment]
@sohamjadhav95 Thanks. Your checks are failing. Converting this to draft till they are fixed.

**Comment by ankurankan** [comment]
@sohamjadhav95 Your doctests are failing.

**Comment by sohamjadhav95** [comment]
@ankurankan I’ll reproduce the doctest failures locally and fix them at earliest.

**Comment by sohamjadhav95** [comment]
All Checks are passing now, let me know if there are any adjustments.

File: pgmpy/models/LinearGaussianBayesianNetwork.py [hunk 6]
```diff
@@ -974,6 +974,119 @@ def fit(
...
+        # Step 2: Check that fit() was called before fit_update()
+        if len(self.get_cpds()) == 0:
+            raise ValueError(
+                "fit_update() requires the model to be first fitted using fit(). Please call fit() before fit_update()."
+            )
+
+        if n_prev_samples is None:
+            n_prev_samples = data.shape[0]
+
+        # Step 3: Get topological order — used for joint stats and CPD extraction
```
Comment (ankurankan) [inline pgmpy/models/LinearGaussianBayesianNetwork.py]: Please remove UTF-8 characters.

File: pgmpy/models/LinearGaussianBayesianNetwork.py [hunk 7]
```diff
@@ -974,6 +974,119 @@ def fit(
...
+    def fit_update(
+        self,
+        data: pd.DataFrame,
+        n_prev_samples: int | None = None,
+    ) -> LinearGaussianBayesianNetwork:
+        """
+        Updates the parameters of the LinearGaussianBayesianNetwork with new
+        data without refitting from scratch.
+
+        Internally, retrieves the joint Gaussian distribution implied by the
```
Comment (ankurankan) [inline pgmpy/models/LinearGaussianBayesianNetwork.py]: Can you include the mathematical formulat of how update works here?

**Review by ankurankan — CHANGES_REQUESTED** [review]

**Comment by sohamjadhav95** [comment]
Ok @ankurankan I have removed UTF-8 characters, also added a notes block includes mathematical formulate of how update works.

File: pgmpy/models/LinearGaussianBayesianNetwork.py [hunk 8]
```diff
@@ -974,6 +974,149 @@ def fit(
...
+            std = sqrt(Sigma[i, i] * n_total / (n_total - 1))
+
+        For a non-root node i with parent indices pa:
+            beta_coeffs = solve(Sigma[pa, pa], Sigma[i, pa])
+            beta_0 = mu[i] - beta_coeffs @ mu[pa]
+            sigma2 = Sigma[i, i] - Sigma[i, pa] @ beta_coeffs
+            std = sqrt(max(sigma2, 0) * n_total / (n_total - k))
+
+        where k = 1 + len(parents). The n_total / (n_total - k) factor matches
+        pgmpy's unbiased std estimator used in ``fit()``.
```
Comment (ankurankan) [inline pgmpy/models/LinearGaussianBayesianNetwork.py]: Should be in the main docstring above, not in notes. No need for a notes section.

Comment (ankurankan) [reply pgmpy/models/LinearGaussianBayesianNetwork.py]: And use latex for math symbols.

**Review by ankurankan — CHANGES_REQUESTED** [review]

**Comment by sohamjadhav95** [comment]
@ankurankan Used latex for math symbols and refined the docstring.

**Comment by sohamjadhav95** [comment]
Hey @ankurankan 
Changes done, waiting for review.

**Comment by sohamjadhav95** [comment]
Hey @ankurankan, I checked both failing tests. The `test_pillai_tests_approx` test doesn't exist in my local dev branch at all, it was added to upstream dev after my fork, so it's a pre-existing failure unrelated to my changes. The sao2.csv linting issue is also a file I never touched. No failure is caused by my PR. All checks related to my changes are passing.

**Comment by sohamjadhav95** [comment]
Hey @ankurankan, all checks are passing now. The linting failure was a missing newline at the end of sao2.csv which came in during the upstream dev merge.

File: .github/PULL_REQUEST_TEMPLATE.md
Comment (ankurankan) [inline .github/PULL_REQUEST_TEMPLATE.md:1]: Why is there change in this file?

File: pgmpy/models/LinearGaussianBayesianNetwork.py
Comment (ankurankan) [inline pgmpy/models/LinearGaussianBayesianNetwork.py:1]: We use a 120 char limit for the code line. Please use that, it helps with readability.

File: pgmpy/models/LinearGaussianBayesianNetwork.py [hunk 9]
```diff
@@ -974,6 +974,155 @@ def fit(
...
+        r"""
+        Updates the parameters of the LinearGaussianBayesianNetwork with new
+        data without refitting from scratch.
+
+        The previous joint Gaussian :math:`(\mu_1, \Sigma_1)` is recovered from
+        the current CPD parameters via ``to_joint_gaussian()``. The new batch
+        statistics :math:`(\mu_2, \Sigma_2)` are computed from the new data with
+        ``ddof=0``. Given :math:`n_1` = ``n_prev_samples``, :math:`n_2` = ``len(data)``,
+        and :math:`n_{total} = n_1 + n_2`, the joint is updated using the exact
+        pooled formulas:
```
Comment (ankurankan) [inline pgmpy/models/LinearGaussianBayesianNetwork.py:994]: Try to make this description explanatory in isolation. It shouldn't be referencing to other methods like `to_joint_gaussian`.

File: pgmpy/models/LinearGaussianBayesianNetwork.py [hunk 10]
```diff
@@ -974,6 +974,155 @@ def fit(
...
+
+        Parameters
+        ----------
+        data : pd.DataFrame
+            New observations to update the model with. Must contain all
+            model variables as columns.
+
+        n_prev_samples : int (optional)
+            The number of samples the model was previously trained on.
+            If None, defaults to the number of rows in the new data.
```
Comment (ankurankan) [inline pgmpy/models/LinearGaussianBayesianNetwork.py:1041]: In the description, please add that this parameter can also be used to control the weight provided to the  older data vs new data. When `size(n_prev_samples) == size(data)`, it is equivalent to specifying equal weights. Increasing would lead to lower weight for new data, etc.

File: asia.net
Comment (ankurankan) [inline asia.net:1]: Why is this file in the PR?

**Review by ankurankan — CHANGES_REQUESTED** [review]
I have added a few comments. I believe there are many unnecessary files in the PR, which will need to be removed.

**Comment by sohamjadhav95** [comment]
Hey @ankurankan, 
I've addressed all the review comments. Apologies for the delay. I've been occupied with my GSoC project.

Changes made:
- Removed the reference to `to_joint_gaussian()` from the docstring and made the description self-contained
- Added weight control explanation to `n_prev_samples` description
- Removed unnecessary files that came in during upstream merges
- Reverted the accidental change to the PR template
- Verified all lines are within the 120 char limit

All 24 CI checks are passing. Ready for review.

**Comment by sohamjadhav95** [comment]
Hey @ankurankan,
Just following up for feedback on the work done.

**Comment by sohamjadhav95** [comment]
@ankurankan, The PR has been pending for a while. We have made much progress on the issue. So I'll appreciate any updates or next steps to move it forward.
